libraries/opensk/src/ctap/mod.rs

```diff
@@ -255,120 +257,80 @@ fn truncate_to_char_boundary(s: &str, mut max: usize) -> &str {
-                            return Err(UserPresenceError::Canceled);
-                        } else {
-                            debug_ctap!(
-                                env,
-                                "Discarded packet with command {} received while sending a KEEPALIVE packet",
-                                cmd,
-                            );
-                        }
-                    }
-                    ProcessedPacket::ContinuationPacket { .. } => {
-                        debug_ctap!(
-                            env,
-                            "Discarded continuation packet received while sending a KEEPALIVE packet",
-                        );
-                    }
-                }
+                debug_ctap!(env, "Timeout sending packet");
             }
+            Ok(SendOrRecvStatus::Sent) => (),
+            _ => panic!("Error sending packet"),
         }
     }
-    Ok(())
 }
 
 /// Blocks for user presence.
 ///
 /// Returns an error in case of timeout, user declining presence request, or keepalive error.
 pub fn check_user_presence<E: Env>(env: &mut E, channel: Channel) -> CtapResult<()> {
     env.user_presence().check_init();
+    let loop_timer = env.clock().make_timer(TOUCH_TIMEOUT_MS);
 
-    // The timeout is N times the keepalive delay.
-    const TIMEOUT_ITERATIONS: usize = TOUCH_TIMEOUT_MS / KEEPALIVE_DELAY_MS;
+    let (cid, transport) = match channel {
+        Channel::MainHid(cid) => (cid, Transport::MainHid),
+        #[cfg(feature = "vendor_hid")]
+        Channel::VendorHid(cid) => (cid, Transport::VendorHid),
+    };
+    let endpoint = transport.usb_endpoint();
 
     // All fallible functions are called without '?' operator to always reach
     // check_complete(...) cleanup function.
 
     let mut result = Err(UserPresenceError::Timeout);
-    for i in 0..=TIMEOUT_ITERATIONS {
-        // First presence check is made without timeout. That way Env implementation may return
-        // user presence check result immediately to client, without sending any keepalive packets.
-        result = env
-            .user_presence()
-            .wait_with_timeout(if i == 0 { 0 } else { KEEPALIVE_DELAY_MS });
-        if !matches!(result, Err(UserPresenceError::Timeout)) {
-            break;
+    while !env.clock().is_elapsed(&loop_timer) {
+        let (status, data) = env.user_presence().wait_with_timeout(KEEPALIVE_DELAY_MS);
+        if let Some((packet, rx_endpoint)) = data {
+            let (received_cid, processed_packet) = CtapHid::<E>::process_single_packet(&packet);
+            if rx_endpoint != endpoint || received_cid != cid {
+                debug_ctap!(
+                    env,
+                    "Received an unrelated packet on endpoint {:?} while sending a KEEPALIVE packet",
+                     rx_endpoint,
+                );
+                let busy_error = CtapHid::<E>::busy_error(received_cid);
+                send_packets(env, rx_endpoint, busy_error);
```

Comment:
We don't continue here, which means we will still process the packet below. Is that expected?

Reply:
Correct and intended. We want to ignore incoming traffic as before, but send a quick response in addition to what we already do.

Reply:
But this is not what we do. If we get a CANCEL on another channel/endpoint, we will process it and cancel the current user presence check.

Reply:
Oh yes, I missed that. Function is split up to allow more `?` usage and simplify the flow.